go/main.go

```diff
@@ -0,0 +1,660 @@
+	}
+}
+
+// readRecords reads nrows records from the file
+func readRecords(head int) ([]Record, error) {
+	return readRecordsFromFile(FileName, head)
+}
+
+// readRecordsFromFile reads nrows records from the file fileName and returns them.
+func readRecordsFromFile(fileName string, head int) ([]Record, error) {
+	if _, err := os.Stat(fileName); os.IsNotExist(err) {
+		createFile()
+	}
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, fmt.Errorf("could not open file: %w", err)
+	}
+	defer file.Close()
+	reader := csv.NewReader(file)
+
+	lines := [][]string{}
+	linesRead := -1
+
+	if head == -1 {
+		// read all
+		lines, err = reader.ReadAll()
+		if err != nil {
+			return nil, fmt.Errorf("could not read CSV: %w", err)
+		}
+	} else {
+		// read n first nrows
+		for i := 0; i < (head + 1); i++ {
+			line, err := reader.Read()
+			lines = append(lines, line)
+			if err != nil {
+				// NOTE: i can happen that the head is greater
+				// thant the number of lines in the file.
+				linesRead = i - 1 // avoid the header
+				break
+			}
+		}
+	}
+
+	var records []Record
+	if head == 0 || linesRead == 0 || len(lines) < 2 {
+		return records, nil
+	}
+	for _, line := range lines[1:] {
+		timestamp, _ := time.Parse(TimeFormat, line[0])
+		records = append(records, Record{timestamp, line[1], line[2]})
+	}
+
+	return records, nil
+}
+
+// checkAction checks in or out.
+func checkAction(filename, notes string) {
+	records, err := readRecordsFromFile(filename, 1)
+	if err != nil {
+		fmt.Println("Error:", err)
+		return
+	}
+
+	var kind string
+	if len(records) == 0 || records[0].Kind == "out" {
+		kind = "in"
+	} else {
+		kind = "out"
+	}
+
+	timestamp := time.Now().Format(TimeFormat)
+	line := fmt.Sprintf("%s,%s,%s", timestamp, kind, notes)
+	if err := writeRecords(filename, line); err != nil {
+		fmt.Println("Error:", err)
+	}
+
+	fmt.Printf("Check %s at %s\n", kind, timestamp)
+}
+
+// writeRecords writes a new line to the file.
+func writeRecords(fileName, newLine string) error {
+	prevFile, err := os.Open(fileName)
+	if err != nil {
+		return err
+	}
+	defer prevFile.Close()
+
+	newFile, err := os.CreateTemp("", "takt_tempfile.csv")
+	if err != nil {
+		fmt.Printf("Error: could not create temp file")
+		return err
+	}
+	defer newFile.Close()
+
+	newWriter := bufio.NewWriter(newFile)
+	defer newWriter.Flush()
+	_, err = newWriter.WriteString(fmt.Sprintf("%s,%s,%s\n", Header[0], Header[1], Header[2]))
+	if err != nil {
+		fmt.Printf("Error: could not write to temp file")
+		return err
+	}
+	_, err = newWriter.WriteString(newLine + "\n")
+	if err != nil {
+		fmt.Printf("Error: could not write to temp file")
+		return err
+	}
+
+	prevReader := bufio.NewReader(prevFile)
+
+	// drop the header
+	_, _, err = prevReader.ReadLine()
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(newWriter, prevReader)
+	if err != nil {
+		return err
+	}
+
+	if err := os.Rename(newFile.Name(), fileName); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
I'm not sure if this is a good implementation. The final goal is to avoid loading the entire file into memory. Additionally, I want (non-negotiable) the file to contain the latest entries at the beginning, and since it's a CSV, this complicates things a bit.

On one hand, I have the CSV file that contains the entire history, and on the other, I create a new file with the latest entry (check in/out) and below it, the content of the original file that contains the entire history, to finally overwrite the old one with this last one.

I have the feeling that this can be improved.